src/main/java/jcifs/smb/Kerb5Authenticator.java

```diff
@@ -0,0 +1,439 @@
+     * @param domain
+     *            domain for NTLM fallback
+     * @param username
+     *            user for NTLM fallback
+     * @param password
+     *            password for NTLM fallback
+     */
+    public Kerb5Authenticator ( Subject subject, String domain, String username, String password ) {
+        super(domain, username, password);
+        this.canFallback = true;
+        this.subject = subject;
+    }
+
+
+    /**
+     * Testing only: force fallback to NTLM
+     * 
+     * @param forceFallback
+     *            the forceFallback to set
+     */
+    public void setForceFallback ( boolean forceFallback ) {
+        this.forceFallback = forceFallback;
+    }
+
+
+    /**
+     * 
+     * {@inheritDoc}
+     *
+     * @see jcifs.smb.NtlmPasswordAuthenticator#createContext(jcifs.CIFSContext, java.lang.String, java.lang.String,
+     *      byte[], boolean)
+     */
+    @Override
+    public SSPContext createContext ( CIFSContext tc, String targetDomain, String host, byte[] initialToken, boolean doSigning ) throws SmbException {
+        if ( host.indexOf('.') < 0 && host.toUpperCase(Locale.ROOT).equals(host) ) {
+            // this is not too good, probably should better pass the address and check that it is a netbios one.
+            // While we could look up the domain controller/KDC we cannot really make the java kerberos implementation
+            // use a KDC of our choice.
+            // A potential workaround would be to try to get the server FQDN by reverse lookup, but this might have
+            // security implications and also is not how Microsoft does it.
+            throw new SmbUnsupportedOperationException("Cannot use netbios/short names with kerberos authentication, have " + host);
+        }
+        try {
+            NegTokenInit tok = new NegTokenInit(initialToken);
+            if ( log.isDebugEnabled() ) {
+                log.debug("Have initial token " + tok);
+            }
+            if ( tok.getMechanisms() != null ) {
+                Set<ASN1ObjectIdentifier> mechs = new HashSet<>(Arrays.asList(tok.getMechanisms()));
+                boolean foundKerberos = false;
+                for ( ASN1ObjectIdentifier mech : Kerb5Context.SUPPORTED_MECHS ) {
+                    foundKerberos |= mechs.contains(mech);
+                }
+                if ( ( !foundKerberos || this.forceFallback ) && this.canFallback && tc.getConfig().isAllowNTLMFallback() ) {
+                    log.debug("Falling back to NTLM authentication");
+                    return super.createContext(tc, targetDomain, host, initialToken, doSigning);
+                }
+                else if ( !foundKerberos ) {
+                    throw new SmbUnsupportedOperationException("Server does not support kerberos authentication");
+                }
+            }
+        }
+        catch ( SmbException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)